src/main/sessions.ts

```diff
@@ -420,12 +422,118 @@ class JupyterLabSession {
             this._window.show();
         });
 
-        this._window.loadURL(url.format({
-            pathname: path.join(__dirname, '../browser/index.html'),
-            protocol: 'file:',
-            slashes: true,
-            search: encodeURIComponent(JSON.stringify(this.info))
-        }));
+        const DESKTOP_APP_ASSETS_PATH = 'desktop-app-assets';
+
+        this._window.loadURL(`http://localhost:${appConfig.jlabPort}/${DESKTOP_APP_ASSETS_PATH}/index.html?${encodeURIComponent(JSON.stringify(this.info))}`);
+
+        const cookies: Map<string, string> = new Map();
+
+        const parseCookie = (cookie: string): Electron.CookiesSetDetails => {
+            const parts = cookie.split(';');
+            const nameVal = parts[0].split('=');
+            const name = nameVal[0].trim();
+            const value = nameVal[1].trim();
+
+            const cookieObj: Electron.CookiesSetDetails = {
+                url: `http://localhost:${appConfig.jlabPort}`,
+                name: name,
+                value: value
+            };
+
+            for (let i = 1; i < parts.length; ++i) {
+                const nameVal = parts[i].split('=');
+                const name = nameVal[0].trim();
+
+                switch (name) {
+                    case 'expires':
+                        {
+                            const value = nameVal[1].trim();
+                            cookieObj.expirationDate = Date.parse(value);
+                        }
+                        break;
+                    case 'Path':
+                        {
+                            const value = nameVal[1].trim();
+                            cookieObj.path = value;
+                        }
+                        break;
+                    case 'HttpOnly':
+                        cookieObj.httpOnly = true;
+                        break;
+                }
+            }
+
+            return cookieObj;
+        };
+
+        this._window.webContents.session.protocol.interceptBufferProtocol("http", (req, callback) => {
+            const desktopAppAssetsPrefix = `http://localhost:${appConfig.jlabPort}/${DESKTOP_APP_ASSETS_PATH}`;
+            const appAssetsDir = path.normalize(path.join(__dirname, '../../'));
+    
+            if (req.url.startsWith(desktopAppAssetsPrefix)) {
+                let assetPath = req.url.substring(desktopAppAssetsPrefix.length + 1);
+                const qMark = assetPath.indexOf('?');
+                if (qMark !== -1) {
+                    assetPath = assetPath.substring(0, qMark);
+                }
+                const assetFilePath = path.join(appAssetsDir, assetPath);
+                const assetContent = fs.readFileSync(assetFilePath);
+                callback(assetContent);
+            } else {
```

Comment:
Could you elaborate on what is happening in the `else` clause? My current understanding is that the `if ()` clase is for hosting static assets and `else` assumes that if it is not a static asset, then this is a server request and modifies cookies and handles file upload, but it is not clear to me _why_ specific parts are needed.

Should we split up the handler code into separate methods to better signal the intent and have easier time in the future?

Reply:
yes, you are right. since interceptBufferProtocol doesn't provide a default callback handler, we are having to make the requests to JupyterLab server in the `else` section. During my experiments, I found out that cookies were not properly set for some of the requests (websocket handshake?). So, to be on the safe side, I am setting them with both `this._window.webContents.session.cookies.set(cookieObject);` and `requestHeaders['Cookie'] = Array.from(cookies.values()).join('; ');` for every request. I am not sure at the moment how we can optimize this.